src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -21,133 +21,505 @@
         if (instance != null)
             mbeanName += ",instance=" + instance;
         subsnitch = snitch;
-        update = new Runnable()
+
+        if (DatabaseDescriptor.isDaemonInitialized())
         {
-            public void run()
+            open();
+        }
+    }
+
+    /**
+     * Update configuration of the background tasks and restart the various scheduler tasks
+     * if the configured rates for these tasks have changed.
+     */
+    public void applyConfigChanges(int newDynamicUpdateInternal, int newDynamicSampleUpdateInterval, double newDynamicBadnessThreshold)
+    {
+        if (DatabaseDescriptor.isDaemonInitialized())
+        {
+            if (dynamicUpdateInterval != newDynamicUpdateInternal || updateScoresScheduler == null)
             {
-                updateScores();
+                cancelAndWait(updateScoresScheduler, Math.max(1, dynamicUpdateInterval), TimeUnit.MILLISECONDS);
+                updateScoresScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateScores, newDynamicUpdateInternal, newDynamicUpdateInternal, TimeUnit.MILLISECONDS);
             }
-        };
-        reset = new Runnable()
+
+            if (dynamicSampleUpdateInterval != newDynamicSampleUpdateInterval || updateSamplesScheduler == null)
+            {
+                cancelAndWait(updateSamplesScheduler, Math.max(1, dynamicSampleUpdateInterval), TimeUnit.MILLISECONDS);
+                if (newDynamicSampleUpdateInterval > 0)
+                    updateSamplesScheduler = ScheduledExecutors.scheduledTasks.scheduleWithFixedDelay(this::updateSamples, newDynamicSampleUpdateInterval, newDynamicSampleUpdateInterval, TimeUnit.MILLISECONDS);
+            }
+        }
+
+        dynamicUpdateInterval = newDynamicUpdateInternal;
+        dynamicSampleUpdateInterval = newDynamicSampleUpdateInterval;
+        dynamicBadnessThreshold = newDynamicBadnessThreshold;
+
+        if (dynamicSampleUpdateInterval > 0)
+            probeRateLimiter.setRate(dynamicSampleUpdateInterval);
+    }
+
+    public synchronized void open()
+    {
+        applyConfigChanges(DatabaseDescriptor.getDynamicUpdateInterval(),
+                           DatabaseDescriptor.getDynamicSampleUpdateInterval(),
+                           DatabaseDescriptor.getDynamicBadnessThreshold());
+
+        MBeanWrapper.instance.registerMBean(this, mbeanName);
+        mbeanRegistered = true;
+    }
+
+    public synchronized void close()
+    {
+        cancelAndWait(updateScoresScheduler, Math.max(1, dynamicUpdateInterval), TimeUnit.MILLISECONDS);
+        cancelAndWait(updateSamplesScheduler, Math.max(1, dynamicSampleUpdateInterval), TimeUnit.MILLISECONDS);
+        updateScoresScheduler = null;
+        updateSamplesScheduler = null;
+
+        for (AnnotatedMeasurement measurement : samples.values())
         {
-            public void run()
+            cancelAndWait(measurement.probeFuture, PING.getTimeout(), TimeUnit.MILLISECONDS);
```

Comment:
Yea, I suppose that the probe could at most be waiting for the rate limiter interval (in fact it will be since the ratelimiter uses `sleepUninterruptibly` afaik. Per your above comments I removed the timeouts and just wait for it.

Also I reworked the cancellation method to take a list of futures so that we can cancel all of them and then wait on all of them.